audisp-tacplus.c

```diff
@@ -93,190 +107,53 @@ hup_handler(int sig __attribute__ ((unused)))
-                        server_buf, gai_strerror(rv));
-                }
-            }
-            else {
-                syslog(LOG_ERR, "maximum number of servers (%d) exceeded, "
-                    "skipping", TAC_PLUS_MAXSERVERS);
-            }
-        }
-        else if(debug) /* ignore unrecognized lines, unless debug on */
-            syslog(LOG_WARNING, "%s: unrecognized parameter: %s",
-                progname, lbuf);
-    }
-
-    if(level == 0 && (!tac_service[0] || tac_srv_no == 0))
-        syslog(LOG_ERR, "%s version %d.%d.%d: missing tacacs fields in file %s, %d servers",
-            progname, _VMAJ, _VMIN, _VPATCH, configfile, tac_srv_no);
-
-    if(debug) {
-        int n;
-        syslog(LOG_NOTICE, "%s version %d.%d.%d tacacs service=%s", progname,
-            _VMAJ, _VMIN, _VPATCH, tac_service);
+    hup = 0;
 
-        for(n = 0; n < tac_srv_no; n++)
-            syslog(LOG_DEBUG, "%s: tacacs server[%d] { addr=%s, key='%s' }",
-                progname, n, tac_ntop(tac_srv[n].addr->ai_addr),
-                tac_srv[n].key);
-    }
+    connected_ok = 0; /*  reset connected state (for possible vrf) */
 
-    fclose(conf);
+    /* load config file: tacacs_config_file */
+    tacacs_ctrl = parse_config_file(tacacs_config_file);
 }
 
-
-static void
-reload_config(void)
+/*
+ * Get user name by UID, and return NULL when not found user name by UID.
+ * The returned username should be free by caller.
+ */
+char *lookup_logname(uid_t auid)
 {
-    int i, nservers;
-
-    hup = 0;
-
-    /*  reset the config variables that we use, freeing memory where needed */
-    nservers = tac_srv_no;
-    tac_srv_no = 0;
-    tac_key_no = 0;
-    vrfname[0] = '\0';
-    tac_service[0] = '\0';
-    tac_protocol[0] = '\0';
-    tac_login[0] = '\0';
-    debug = 0;
-    acct_all = 0;
-    tac_timeout = 0;
-
-    for(i = 0; i < nservers; i++) {
-        if(tac_srv[i].key) {
-            free(tac_srv[i].key);
-            tac_srv[i].key = NULL;
-        }
-        tac_srv[i].addr = NULL;
+    struct passwd *pws;
+    pws = getpwuid(auid);
+    if (pws == NULL) {
+        /* Failed to get user information. */
+        return NULL;
     }
-
-    connected_ok = 0; /*  reset connected state (for possible vrf) */
-
-    audisp_tacplus_config(configfile, 0);
+    
+    int new_buffer_size = strlen(pws->pw_name) + 1;
+    char* username = malloc(new_buffer_size);
+    if (username == NULL) {
+        /* Failed to allocate new buffer. */
+        return NULL;
+    }
+    
+    memset(username, 0, new_buffer_size);
+    strncpy(username, pws->pw_name, new_buffer_size-1);
+    return username;
 }
 
 int
 main(int argc, char *argv[])
 {
 	char tmp[MAX_AUDIT_MESSAGE_LENGTH+1];
 	struct sigaction sa;
+    
+    /* initialize random seed*/
+    srand(time(NULL));
 
-    /* if there is an argument, it is an alternate configuration file */
-    if(argc > 1)
-        configfile = argv[1];
```

Comment:
>configfile [](http://example.com/codeflow?start=8&length=10)

Original the command line argument support override the configfile, and you are just using a "/etc/tacplus_nss.conf". Possible to leverage this argument?

Reply:
Fixed, now using audisp-tacplus.conf to config the "/etc/tacplus_nss.conf"